File: compatibility/python/smoke.py
import os
import sys
import time

from confluent_kafka import Consumer, Producer, TopicPartition
from confluent_kafka.admin import AdminClient, NewTopic


bootstrap = os.environ.get("CASCADE_BOOTSTRAP_SERVERS", "127.0.0.1:19092")
topic = sys.argv[1] if len(sys.argv) > 1 else f"python-compat-{time.time_ns()}"
expected = [f"python-{index}" for index in range(25)]

admin = AdminClient({"bootstrap.servers": bootstrap, "socket.timeout.ms": 10_000})
creation = admin.create_topics([NewTopic(topic, num_partitions=1, replication_factor=1)])[topic]
creation.result(15)

delivery_errors = []


def delivered(error, _message):
    if error is not None:
        delivery_errors.append(error)


producer = Producer(
    {
        "bootstrap.servers": bootstrap,
        "enable.idempotence": True,
        "acks": "all",
        "message.timeout.ms": 15_000,
    }
)
for value in expected:
    producer.produce(topic, value=value.encode("utf-8"), on_delivery=delivered)
remaining = producer.flush(20)
if remaining or delivery_errors:
    raise RuntimeError(f"Python producer failed: remaining={remaining}, errors={delivery_errors}")

consumer = Consumer(
    {
        "bootstrap.servers": bootstrap,
        "group.id": f"{topic}-reader",
        "enable.auto.commit": False,
        "auto.offset.reset": "earliest",
    }
)
actual = []
try:
    consumer.assign([TopicPartition(topic, 0, 0)])
    deadline = time.monotonic() + 20
    while len(actual) < len(expected) and time.monotonic() < deadline:
        message = consumer.poll(0.25)
        if message is None:
            continue
        if message.error():
            raise RuntimeError(message.error())
        actual.append(message.value().decode("utf-8"))
finally:
    consumer.close()

if actual != expected:
    raise RuntimeError(f"Python client mismatch: expected={expected}, actual={actual}")
print(f"Python confluent-kafka verified {len(actual)}/{len(expected)} records through {bootstrap}")

File: compatibility/node/smoke.js
const { Kafka, logLevel } = require('kafkajs')

const bootstrap = process.env.CASCADE_BOOTSTRAP_SERVERS || '127.0.0.1:19092'
const topic = process.argv[2] || `node-compat-${Date.now()}`
const replicationFactor = Number(process.env.CASCADE_REPLICATION_FACTOR || '1')
if (!Number.isInteger(replicationFactor) || replicationFactor < 1 || replicationFactor > 32767) {
  throw new Error('CASCADE_REPLICATION_FACTOR must be an integer from 1 to 32767')
}
const expected = Array.from({ length: 25 }, (_, index) => `node-${index}`)

async function main() {
  const kafka = new Kafka({
    clientId: 'cascade-kafkajs-compatibility',
    brokers: bootstrap.split(','),
    connectionTimeout: 5000,
    requestTimeout: 15000,
    logLevel: logLevel.NOTHING
  })
  const admin = kafka.admin()
  const producer = kafka.producer({ idempotent: true, maxInFlightRequests: 1 })
  const consumer = kafka.consumer({ groupId: `${topic}-readers` })

  await admin.connect()
  try {
    await admin.createTopics({
      waitForLeaders: true,
      topics: [{ topic, numPartitions: 1, replicationFactor }]
    })
  } finally {
    await admin.disconnect()
  }

  await producer.connect()
  try {
    await producer.send({
      topic,
      acks: -1,
      messages: expected.map(value => ({ value }))
    })
  } finally {
    await producer.disconnect()
  }

  const actual = []
  await consumer.connect()
  try {
    await consumer.subscribe({ topic, fromBeginning: true })
    await new Promise(async (resolve, reject) => {
      const timeout = setTimeout(() => reject(new Error(`KafkaJS received only ${actual.length}/${expected.length} records`)), 20000)
      try {
        await consumer.run({
          eachMessage: async ({ message }) => {
            actual.push(message.value.toString('utf8'))
            if (actual.length === expected.length) {
              clearTimeout(timeout)
              resolve()
            }
          }
        })
      } catch (error) {
        clearTimeout(timeout)
        reject(error)
      }
    })
    await consumer.commitOffsets([{ topic, partition: 0, offset: String(expected.length) }])
  } finally {
    await consumer.disconnect()
  }

  if (JSON.stringify(actual) !== JSON.stringify(expected)) {
    throw new Error(`KafkaJS mismatch: expected=${JSON.stringify(expected)}, actual=${JSON.stringify(actual)}`)
  }
  await admin.connect()
  try {
    const offsets = await admin.fetchOffsets({ groupId: `${topic}-readers`, topics: [topic] })
    const committed = offsets.find(item => item.topic === topic)?.partitions.find(item => item.partition === 0)?.offset
    if (committed !== String(expected.length)) {
      throw new Error(`KafkaJS committed offset mismatch: expected=${expected.length}, actual=${committed}`)
    }
  } finally {
    await admin.disconnect()
  }
  console.log(`KafkaJS verified ${actual.length}/${expected.length} records through ${bootstrap}`)
}

main().catch(error => {
  console.error(error)
  process.exitCode = 1
})
